File: javaspider/src/main/java/com/javaspider/examples/ecommerce/UniversalEcommerceDetector.java
package com.javaspider.examples.ecommerce;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Pattern;

public final class UniversalEcommerceDetector {
    private UniversalEcommerceDetector() {
    }

    public static final class DetectionResult {
        public boolean isEcommerce = false;
        public double confidence = 0.0;
        public String siteFamily = "generic";
        public String platform = "";
        public List<String> detectedFeatures = new ArrayList<>();
        public String currency = "";
        public boolean hasJsonLd = false;
        public boolean hasNextData = false;
        public boolean hasInitialState = false;
        public boolean priceApiDetected = false;
        public String cartUrl = "";
        public List<String> categoryUrls = new ArrayList<>();

        public Map<String, Object> toMap() {
            Map<String, Object> result = new LinkedHashMap<>();
            result.put("is_ecommerce", isEcommerce);
            result.put("confidence", confidence);
            result.put("site_family", siteFamily);
            result.put("platform", platform);
            result.put("detected_features", detectedFeatures);
            result.put("currency", currency);
            result.put("has_jsonld", hasJsonLd);
            result.put("has_next_data", hasNextData);
            result.put("has_initial_state", hasInitialState);
            result.put("price_api_detected", priceApiDetected);
            result.put("cart_url", cartUrl);
            result.put("category_urls", categoryUrls);
            return result;
        }
    }

    private record UrlSignature(List<String> patterns, double confidence, String currency) {
    }

    private record PlatformSignature(List<String> needles, double confidence) {
    }

    private static final Map<String, UrlSignature> URL_SIGNATURES = new LinkedHashMap<>();
    private static final Map<String, PlatformSignature> PLATFORM_SIGNATURES = new LinkedHashMap<>();
    private static final String[] HTML_SIGNALS = {
        "\"@type\"\\s*:\\s*\"Product\"",
        "\"@type\"\\s*:\\s*\"Offer\"",
        "\"@type\"\\s*:\\s*\"AggregateOffer\"",
        "class=[\"'][^\"']*price[^\"']*[\"']",
        "class=[\"'][^\"']*product[^\"']*[\"']",
        "class=[\"'][^\"']*add-to-cart[^\"']*[\"']",
        "shopping[\\-]?cart",
        "data-product-id",
        "data-sku",
        "data-variant-id",
        "itemtype=[\"']https?://schema\\.org/Product[\"']",
        "[\\$€£¥₹₩₽][\\d,]+\\.?\\d*"
    };

    static {
        putUrl("jd", 0.95, "CNY", "jd\\.com", "jd\\.hk");
        putUrl("taobao", 0.95, "CNY", "taobao\\.com");
        putUrl("tmall", 0.95, "CNY", "tmall\\.com");
        putUrl("pinduoduo", 0.95, "CNY", "pinduoduo\\.com", "yangkeduo\\.com", "pdd\\.com");
        putUrl("1688", 0.95, "CNY", "1688\\.com");
        putUrl("suning", 0.90, "CNY", "suning\\.com");
        putUrl("vip", 0.95, "CNY", "vip\\.com", "vipshop\\.com");
        putUrl("xiaohongshu", 0.95, "CNY", "xiaohongshu\\.com", "xhscdn\\.com");
        putUrl("douyin-shop", 0.90, "CNY", "douyin\\.com", "jinritemai\\.com", "life\\.douyin");
        putUrl("kuaishou-shop", 0.90, "CNY", "kuaishou\\.com", "kwai\\.com");
        putUrl("amazon", 0.95, "USD", "amazon\\.(com|co\\.uk|de|fr|it|es|co\\.jp|com\\.au|ca|in|com\\.br|com\\.mx|nl|pl|se|ae|sa|sg)");
        putUrl("ebay", 0.95, "USD", "ebay\\.(com|co\\.uk|de|fr|it|es|com\\.au|ca|ch|at|be|nl|ie|pl|ph|in|my|sg)");
        putUrl("aliexpress", 0.95, "USD", "aliexpress\\.(com|us|es|ru|pt|fr|de|it|nl|ja|ko|ar|th|vi|id|he|pl|tr)");
        putUrl("lazada", 0.95, "USD", "lazada\\.(com|com\\.my|com\\.ph|co\\.id|co\\.th|vn)");
        putUrl("shopee", 0.95, "USD", "shopee\\.(com|co\\.th|co\\.id|com\\.my|com\\.ph|vn|tw|br|cl|pl)");
        putUrl("walmart", 0.95, "USD", "walmart\\.(com|ca)");
        putUrl("target", 0.95, "USD", "target\\.com");
        putUrl("temu", 0.95, "USD", "temu\\.(com|co\\.uk|co\\.jp|de|fr|es|it|nl|be|at|se|pl|pt|ch|dk|fi|gr|cz|hu|ro|bg)");
        putUrl("shein", 0.95, "USD", "shein\\.(com|co\\.uk|co\\.jp|de|fr|es|it|nl|se|at|pl|pt|ch|dk|fi|gr|cz|hu|ro|bg)");
        putUrl("mercadolibre", 0.95, "USD", "mercadolibre\\.(com\\.ar|com\\.mx|com\\.br|com\\.co|com\\.pe|cl|com\\.uy)");
        putUrl("ozon", 0.95, "RUB", "ozon\\.ru");
        putUrl("wildberries", 0.95, "RUB", "wildberries\\.ru");
        putUrl("allegro", 0.95, "PLN", "allegro\\.(pl|cz|sk|hu|ro|bg)");

        putPlatform("shopify", 0.90, "shopify.com", "cdn.shopify", "Shopify.theme");
        putPlatform("magento", 0.90, "Magento_", "mage-cache", "mage/cookies");
        putPlatform("woocommerce", 0.85, "woocommerce", "wp-content/plugins/woocommerce");
        putPlatform("bigcommerce", 0.90, "bigcommerce", "bc.js");
        putPlatform("prestashop", 0.90, "prestashop", "ps-shoppingcart");
        putPlatform("wix", 0.80, "wix.com", "wixstores");
        putPlatform("salesforce", 0.85, "demandware", "sfcc", "commercecloud");
    }

    public static DetectionResult detect(String url, String html) {
        DetectionResult result = new DetectionResult();
        if (url == null || url.isBlank()) {
            return result;
        }
        merge(result, detectByUrl(url));
        merge(result, detectByPlatform(html));
        merge(result, detectByHtmlSignals(html));
        merge(result, detectByStructuredData(html));
        merge(result, detectByPriceApi(html));
        applyFamilyUrls(result);
        return result;
    }

    private static void putUrl(String family, double confidence, String currency, String... patterns) {
        URL_SIGNATURES.put(family, new UrlSignature(List.of(patterns), confidence, currency));
    }

    private static void putPlatform(String platform, double confidence, String... needles) {
        PLATFORM_SIGNATURES.put(platform, new PlatformSignature(List.of(needles), confidence));
    }

    private static void merge(DetectionResult result, DetectionResult candidate) {
        if (candidate == null) {
            return;
        }
        result.isEcommerce = result.isEcommerce || candidate.isEcommerce;
        result.confidence = Math.max(result.confidence, candidate.confidence);
        if ("generic".equals(result.siteFamily) && !"generic".equals(candidate.siteFamily)) {
            result.siteFamily = candidate.siteFamily;
        }
        if (result.platform.isBlank() && !candidate.platform.isBlank()) {
            result.platform = candidate.platform;
        }
        if (result.currency.isBlank() && !candidate.currency.isBlank()) {
            result.currency = candidate.currency;
        }
        result.hasJsonLd = result.hasJsonLd || candidate.hasJsonLd;
        result.hasNextData = result.hasNextData || candidate.hasNextData;
        result.hasInitialState = result.hasInitialState || candidate.hasInitialState;
        result.priceApiDetected = result.priceApiDetected || candidate.priceApiDetected;
        for (String feature : candidate.detectedFeatures) {
            if (!result.detectedFeatures.contains(feature)) {
                result.detectedFeatures.add(feature);
            }
        }
    }

    private static DetectionResult detectByUrl(String url) {
        for (Map.Entry<String, UrlSignature> entry : URL_SIGNATURES.entrySet()) {
            for (String pattern : entry.getValue().patterns()) {
                if (Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(url).find()) {
                    DetectionResult result = new DetectionResult();
                    result.isEcommerce = true;
                    result.confidence = entry.getValue().confidence();
                    result.siteFamily = entry.getKey();
                    result.platform = entry.getKey();
                    result.currency = entry.getValue().currency();
                    result.detectedFeatures.add("url_pattern");
                    return result;
                }
            }
        }
        return null;
    }

    private static DetectionResult detectByPlatform(String html) {
        if (html == null || html.isBlank()) {
            return null;
        }
        String lowered = html.toLowerCase(Locale.ROOT);
        for (Map.Entry<String, PlatformSignature> entry : PLATFORM_SIGNATURES.entrySet()) {
            for (String needle : entry.getValue().needles()) {
                if (lowered.contains(needle.toLowerCase(Locale.ROOT))) {
                    DetectionResult result = new DetectionResult();
                    result.isEcommerce = true;
                    result.confidence = entry.getValue().confidence();
                    result.platform = entry.getKey();
                    result.detectedFeatures.add("platform_signature");
                    return result;
                }
            }
        }
        return null;
    }

    private static DetectionResult detectByHtmlSignals(String html) {
        if (html == null || html.isBlank()) {
            return null;
        }
        int signals = 0;
        for (String pattern : HTML_SIGNALS) {
            if (Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(html).find()) {
                signals++;
            }
        }
        boolean hasJsonLd = html.contains("application/ld+json");
        boolean hasNextData = html.contains("__NEXT_DATA__") || html.contains("__NUXT__");
        boolean hasInitialState = html.contains("__INITIAL_STATE__") || html.contains("__PRELOADED_STATE__");
        if (hasJsonLd) {
            signals += 2;
        }
        if (hasNextData || hasInitialState) {
            signals++;
        }
        if (signals < 2) {
            return null;
        }
        DetectionResult result = new DetectionResult();
        result.isEcommerce = true;
        result.confidence = Math.min(0.85, signals * 0.15);
        result.hasJsonLd = hasJsonLd;
        result.hasNextData = hasNextData;
        result.hasInitialState = hasInitialState;
        result.detectedFeatures.add("html_structure");
        return result;
    }

    private static DetectionResult detectByStructuredData(String html) {
        if (html == null || html.isBlank()) {
            return null;
        }
        if (Pattern.compile("\"@type\"\\s*:\\s*\"(Product|Offer|AggregateOffer|ItemList)\"", Pattern.CASE_INSENSITIVE).matcher(html).find()
            || (html.contains("\"offers\"") && (html.contains("\"price\"") || html.contains("\"priceCurrency\"")))) {
            DetectionResult result = new DetectionResult();
            result.isEcommerce = true;
            result.confidence = 0.75;
            result.hasJsonLd = html.contains("application/ld+json");
            result.detectedFeatures.add("structured_product_data");
            return result;
        }
        return null;
    }

    private static DetectionResult detectByPriceApi(String html) {
        if (html == null || html.isBlank()) {
            return null;
        }
        String lowered = html.toLowerCase(Locale.ROOT);
        int hits = 0;
        for (String needle : List.of("price", "amount", "currency", "discount", "saleprice", "stock")) {
            if (lowered.contains(needle)) {
                hits++;
            }
        }
        if (hits < 3) {
            return null;
        }
        DetectionResult result = new DetectionResult();
        result.isEcommerce = true;
        result.confidence = 0.50;
        result.priceApiDetected = true;
        result.detectedFeatures.add("price_api");
        return result;
    }

    private static void applyFamilyUrls(DetectionResult result) {
        switch (result.siteFamily) {
            case "jd" -> {
                result.cartUrl = "https://cart.jd.com/";
                result.categoryUrls = List.of("https://channel.jd.com/");
            }
            case "taobao" -> {
                result.cartUrl = "https://cart.taobao.com/";
                result.categoryUrls = List.of("https://www.taobao.com/tbhome/");
            }
            case "tmall" -> {
                result.cartUrl = "https://cart.tmall.com/";
                result.categoryUrls = List.of("https://www.tmall.com/");
            }
            case "amazon" -> {
                result.cartUrl = "https://www.amazon.com/gp/cart/";
                result.categoryUrls = List.of("https://www.amazon.com/best-sellers/");
            }
            default -> {
            }
        }
    }
}

File: pyspider/examples/universal_ecommerce_detector.py
"""Universal e-commerce site detector for pyspider examples."""

import json
import re
from dataclasses import dataclass, field
from typing import Any


@dataclass
class EcommerceDetectionResult:
    is_ecommerce: bool = False
    confidence: float = 0.0
    site_family: str = "generic"
    platform: str = ""
    detected_features: list[str] = field(default_factory=list)
    currency: str = ""
    language: str = ""
    has_jsonld: bool = False
    has_next_data: bool = False
    has_initial_state: bool = False
    price_api_detected: bool = False
    cart_url: str = ""
    category_urls: list[str] = field(default_factory=list)


URL_SIGNATURES: dict[str, dict[str, Any]] = {
    "jd": {"patterns": [r"jd\.com", r"jd\.hk"], "confidence": 0.95, "currency": "CNY"},
    "taobao": {"patterns": [r"taobao\.com"], "confidence": 0.95, "currency": "CNY"},
    "tmall": {"patterns": [r"tmall\.com"], "confidence": 0.95, "currency": "CNY"},
    "pinduoduo": {"patterns": [r"pinduoduo\.com", r"yangkeduo\.com", r"pdd\.com"], "confidence": 0.95, "currency": "CNY"},
    "1688": {"patterns": [r"1688\.com"], "confidence": 0.95, "currency": "CNY"},
    "suning": {"patterns": [r"suning\.com"], "confidence": 0.90, "currency": "CNY"},
    "vip": {"patterns": [r"vip\.com", r"vipshop\.com"], "confidence": 0.95, "currency": "CNY"},
    "xiaohongshu": {"patterns": [r"xiaohongshu\.com", r"xhscdn\.com"], "confidence": 0.95, "currency": "CNY"},
    "douyin-shop": {"patterns": [r"douyin\.com", r"jinritemai\.com", r"life\.douyin"], "confidence": 0.90, "currency": "CNY"},
    "kuaishou-shop": {"patterns": [r"kuaishou\.com", r"kwai\.com"], "confidence": 0.90, "currency": "CNY"},
    "amazon": {"patterns": [r"amazon\.(com|co\.uk|de|fr|it|es|co\.jp|com\.au|ca|in|com\.br|com\.mx|nl|pl|se|ae|sa|sg)"], "confidence": 0.95, "currency": "USD"},
    "ebay": {"patterns": [r"ebay\.(com|co\.uk|de|fr|it|es|com\.au|ca|ch|at|be|nl|ie|pl|ph|in|my|sg)"], "confidence": 0.95, "currency": "USD"},
    "aliexpress": {"patterns": [r"aliexpress\.(com|us|es|ru|pt|fr|de|it|nl|ja|ko|ar|th|vi|id|he|pl|tr)"], "confidence": 0.95, "currency": "USD"},
    "lazada": {"patterns": [r"lazada\.(com|com\.my|com\.ph|co\.id|co\.th|vn)"], "confidence": 0.95, "currency": "USD"},
    "shopee": {"patterns": [r"shopee\.(com|co\.th|co\.id|com\.my|com\.ph|vn|tw|br|cl|pl)"], "confidence": 0.95, "currency": "USD"},
    "rakuten": {"patterns": [r"rakuten\.(co\.jp|com|com\.tw|co\.kr)"], "confidence": 0.95, "currency": "JPY"},
    "walmart": {"patterns": [r"walmart\.(com|ca)"], "confidence": 0.95, "currency": "USD"},
    "bestbuy": {"patterns": [r"bestbuy\.(com|ca)"], "confidence": 0.95, "currency": "USD"},
    "target": {"patterns": [r"target\.com"], "confidence": 0.95, "currency": "USD"},
    "costco": {"patterns": [r"costco\.(com|ca|co\.uk|com\.au|com\.jp|com\.kr|com\.tw|com\.mx)"], "confidence": 0.95, "currency": "USD"},
    "newegg": {"patterns": [r"newegg\.(com|ca|com\.global)"], "confidence": 0.95, "currency": "USD"},
    "temu": {"patterns": [r"temu\.(com|co\.uk|co\.jp|de|fr|es|it|nl|be|at|se|pl|pt|ch|dk|fi|gr|cz|hu|ro|bg)"], "confidence": 0.95, "currency": "USD"},
    "shein": {"patterns": [r"shein\.(com|co\.uk|co\.jp|de|fr|es|it|nl|se|at|pl|pt|ch|dk|fi|gr|cz|hu|ro|bg)"], "confidence": 0.95, "currency": "USD"},
    "mercadolibre": {"patterns": [r"mercadolibre\.(com\.ar|com\.mx|com\.br|com\.co|com\.pe|cl|com\.uy)"], "confidence": 0.95, "currency": "USD"},
    "ozon": {"patterns": [r"ozon\.ru"], "confidence": 0.95, "currency": "RUB"},
    "wildberries": {"patterns": [r"wildberries\.ru"], "confidence": 0.95, "currency": "RUB"},
    "allegro": {"patterns": [r"allegro\.(pl|cz|sk|hu|ro|bg)"], "confidence": 0.95, "currency": "PLN"},
    "cdiscount": {"patterns": [r"cdiscount\.com"], "confidence": 0.95, "currency": "EUR"},
}

PLATFORM_SIGNATURES: dict[str, dict[str, Any]] = {
    "shopify": {"html": ["shopify.com", "cdn.shopify", "Shopify.theme"], "confidence": 0.90},
    "magento": {"html": ["Magento_", "mage-cache", "mage/cookies"], "confidence": 0.90},
    "woocommerce": {"html": ["woocommerce", "wp-content/plugins/woocommerce"], "confidence": 0.85},
    "bigcommerce": {"html": ["bigcommerce", "bc.js"], "confidence": 0.90},
    "prestashop": {"html": ["prestashop", "ps-shoppingcart"], "confidence": 0.90},
    "wix": {"html": ["wix.com", "wixstores"], "confidence": 0.80},
    "squarespace": {"html": ["squarespace.com", "sqs-shop"], "confidence": 0.80},
    "salesforce": {"html": ["demandware", "sfcc", "commercecloud"], "confidence": 0.85},
}

ECOMMERCE_HTML_SIGNALS = [
    r'"@type"\s*:\s*"Product"',
    r'"@type"\s*:\s*"Offer"',
    r'"@type"\s*:\s*"AggregateOffer"',
    r'class=["\'][^"\']*price[^"\']*["\']',
    r'class=["\'][^"\']*product[^"\']*["\']',
    r'class=["\'][^"\']*add-to-cart[^"\']*["\']',
    r'shopping[\-]?cart',
    r'data-product-id',
    r'data-sku',
    r'data-variant-id',
    r'itemtype=["\']https?://schema\.org/Product["\']',
    r'[\$€£¥₹₩₽][\d,]+\.?\d*',
]

SITE_FAMILY_URLS = {
    "jd": {"cart_url": "https://cart.jd.com/", "category_urls": ["https://channel.jd.com/"]},
    "taobao": {"cart_url": "https://cart.taobao.com/", "category_urls": ["https://www.taobao.com/tbhome/"]},
    "tmall": {"cart_url": "https://cart.tmall.com/", "category_urls": ["https://www.tmall.com/"]},
    "amazon": {"cart_url": "https://www.amazon.com/gp/cart/", "category_urls": ["https://www.amazon.com/best-sellers/"]},
}


class UniversalEcommerceDetector:
    @staticmethod
    def detect(url: str, html: str = "", headers: dict[str, str] | None = None) -> EcommerceDetectionResult:
        result = EcommerceDetectionResult()
        if not url:
            return result

        for candidate in (
            UniversalEcommerceDetector._detect_by_url(url),
            UniversalEcommerceDetector._detect_by_platform(html),
            UniversalEcommerceDetector._detect_by_html(html),
            UniversalEcommerceDetector._detect_by_structured_data(html),
            UniversalEcommerceDetector._detect_by_price_api(html, headers or {}),
        ):
            UniversalEcommerceDetector._merge(result, candidate)

        if result.site_family in SITE_FAMILY_URLS:
            profile = SITE_FAMILY_URLS[result.site_family]
            result.cart_url = profile["cart_url"]
            result.category_urls = list(profile["category_urls"])
        return result

    @staticmethod
    def _merge(result: EcommerceDetectionResult, candidate: EcommerceDetectionResult | None) -> None:
        if candidate is None:
            return
        result.is_ecommerce = result.is_ecommerce or candidate.is_ecommerce
        result.confidence = max(result.confidence, candidate.confidence)
        if result.site_family == "generic" and candidate.site_family != "generic":
            result.site_family = candidate.site_family
        if not result.platform and candidate.platform:
            result.platform = candidate.platform
        if not result.currency and candidate.currency:
            result.currency = candidate.currency
        result.has_jsonld = result.has_jsonld or candidate.has_jsonld
        result.has_next_data = result.has_next_data or candidate.has_next_data
        result.has_initial_state = result.has_initial_state or candidate.has_initial_state
        result.price_api_detected = result.price_api_detected or candidate.price_api_detected
        for feature in candidate.detected_features:
            if feature not in result.detected_features:
                result.detected_features.append(feature)

    @staticmethod
    def _detect_by_url(url: str) -> EcommerceDetectionResult | None:
        for family, signature in URL_SIGNATURES.items():
            if any(re.search(pattern, url, re.I) for pattern in signature["patterns"]):
                return EcommerceDetectionResult(
                    is_ecommerce=True,
                    confidence=float(signature["confidence"]),
                    site_family=family,
                    platform=family,
                    currency=str(signature.get("currency", "")),
                    detected_features=["url_pattern"],
                )
        return None

    @staticmethod
    def _detect_by_platform(html: str) -> EcommerceDetectionResult | None:
        if not html:
            return None
        lowered = html.lower()
        for platform, signature in PLATFORM_SIGNATURES.items():
            if any(needle.lower() in lowered for needle in signature["html"]):
                return EcommerceDetectionResult(
                    is_ecommerce=True,
                    confidence=float(signature["confidence"]),
                    platform=platform,
                    detected_features=["platform_signature"],
                )
        return None

    @staticmethod
    def _detect_by_html(html: str) -> EcommerceDetectionResult | None:
        if not html:
            return None
        signals = sum(1 for pattern in ECOMMERCE_HTML_SIGNALS if re.search(pattern, html, re.I))
        has_jsonld = "application/ld+json" in html
        has_next_data = "__NEXT_DATA__" in html or "__NUXT__" in html
        has_initial_state = "__INITIAL_STATE__" in html or "__PRELOADED_STATE__" in html
        if has_jsonld:
            signals += 2
        if has_next_data or has_initial_state:
            signals += 1
        if signals < 2:
            return None
        return EcommerceDetectionResult(
            is_ecommerce=True,
            confidence=min(0.85, signals * 0.15),
            has_jsonld=has_jsonld,
            has_next_data=has_next_data,
            has_initial_state=has_initial_state,
            detected_features=["html_structure"],
        )

    @staticmethod
    def _detect_by_structured_data(html: str) -> EcommerceDetectionResult | None:
        for payload in _json_payloads(html):
            try:
                data = json.loads(payload)
            except json.JSONDecodeError:
                continue
            if _contains_product_shape(data):
                return EcommerceDetectionResult(
                    is_ecommerce=True,
                    confidence=0.75,
                    has_jsonld=True,
                    detected_features=["structured_product_data"],
                )
        return None

    @staticmethod
    def _detect_by_price_api(html: str, headers: dict[str, str]) -> EcommerceDetectionResult | None:
        lowered = html.lower()
        hits = sum(1 for word in ("price", "amount", "currency", "discount", "saleprice", "stock") if word in lowered)
        header_text = " ".join(f"{key}:{value}" for key, value in headers.items()).lower()
        if "json" in header_text and hits >= 2:
            hits += 1
        if hits >= 3:
            return EcommerceDetectionResult(
                is_ecommerce=True,
                confidence=0.50,
                price_api_detected=True,
                detected_features=["price_api"],
            )
        return None


def _json_payloads(html: str) -> list[str]:
    patterns = [
        r'(?is)<script[^>]+type=["\']application/ld\+json["\'][^>]*>(.*?)</script>',
        r'(?is)<script[^>]+type=["\']application/json["\'][^>]*>(.*?)</script>',
        r'(?is)__NEXT_DATA__\s*=\s*(\{.*?\})\s*;',
        r'(?is)__INITIAL_STATE__\s*=\s*(\{.*?\})\s*;',
    ]
    payloads: list[str] = []
    for pattern in patterns:
        payloads.extend(match.group(1).strip() for match in re.finditer(pattern, html))
    return payloads


def _contains_product_shape(value: Any) -> bool:
    if isinstance(value, dict):
        node_type = str(value.get("@type", "")).lower()
        if node_type in {"product", "offer", "aggregateoffer", "itemlist"}:
            return True
        if sum(1 for key in ("sku", "price", "offers", "priceCurrency", "aggregateRating") if key in value) >= 2:
            return True
        return any(_contains_product_shape(child) for child in value.values())
    if isinstance(value, list):
        return any(_contains_product_shape(child) for child in value)
    return False


def detect_ecommerce_site(url: str, html: str = "", headers: dict[str, str] | None = None) -> EcommerceDetectionResult:
    return UniversalEcommerceDetector.detect(url, html, headers)
